plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go

```diff
@@ -0,0 +1,532 @@
+				UpdatesOnly:  c.UpdatesOnly,
+			},
+		},
+	}, nil
+}
+
+// SubscribeGNMI and extract telemetry data
+func (c *CiscoTelemetryGNMI) subscribeGNMI(ctx context.Context, address string, tlscfg *tls.Config, request *gnmi.SubscribeRequest) error {
+	var opt grpc.DialOption
+	if tlscfg != nil {
+		opt = grpc.WithTransportCredentials(credentials.NewTLS(tlscfg))
+	} else {
+		opt = grpc.WithInsecure()
+	}
+
+	client, err := grpc.DialContext(ctx, address, opt)
+	if err != nil {
+		return fmt.Errorf("failed to dial: %v", err)
+	}
+	defer client.Close()
+
+	subscribeClient, err := gnmi.NewGNMIClient(client).Subscribe(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to setup subscription: %v", err)
+	}
+
+	if err = subscribeClient.Send(request); err != nil {
+		return fmt.Errorf("failed to send subscription request: %v", err)
+	}
+
+	log.Printf("D! [inputs.cisco_telemetry_gnmi]: Connection to GNMI device %s established", address)
+	defer log.Printf("D! [inputs.cisco_telemetry_gnmi]: Connection to GNMI device %s closed", address)
+	for ctx.Err() == nil {
+		var reply *gnmi.SubscribeResponse
+		if reply, err = subscribeClient.Recv(); err != nil {
+			if err != io.EOF && ctx.Err() == nil {
+				return fmt.Errorf("aborted GNMI subscription: %v", err)
+			}
+			break
+		}
+
+		c.handleSubscribeResponse(address, reply)
+	}
+	return nil
+}
+
+// HandleSubscribeResponse message from GNMI and parse contained telemetry data
+func (c *CiscoTelemetryGNMI) handleSubscribeResponse(address string, reply *gnmi.SubscribeResponse) {
+	// Check if response is a GNMI Update and if we have a prefix to derive the measurement name
+	response, ok := reply.Response.(*gnmi.SubscribeResponse_Update)
+	if !ok {
+		return
+	}
+
+	var aliasPath string
+	timestamp := time.Unix(0, response.Update.Timestamp)
+	fields := make(map[string]interface{})
+	tags := make(map[string]string)
+
+	var builder bytes.Buffer
+	if response.Update.Prefix != nil {
+		aliasPath = c.handlePath(response.Update.Prefix, tags, &builder)
+	}
+
+	tags["source"], _, _ = net.SplitHostPort(address)
+
+	// Parse individual Update message and create measurements
+	var lastTags map[string]string
+	var lastAliasPath string
+	for _, update := range response.Update.Update {
+		// Prepare tags from prefix
+		currentTags := make(map[string]string, len(tags))
+		for key, val := range tags {
+			currentTags[key] = val
+		}
+		currentAliasPath, currentFields := c.handleTelemetryField(update, currentTags, &builder)
+
+		// Inherent valid alias from prefix parsing
+		if len(aliasPath) > 0 && len(currentAliasPath) == 0 {
+			currentAliasPath = aliasPath
+		}
+
+		// If tags have not changed aggregate fields, if not commit existing before starting anew
+		tagsChanged := len(lastTags) != len(currentTags)
+		for key, val1 := range lastTags {
+			if val2, ok := currentTags[key]; !ok || val2 != val1 {
+				tagsChanged = true
+				break
+			}
+		}
+		if (tagsChanged || currentAliasPath != lastAliasPath) && len(fields) > 0 {
```

Comment:
I think here you will want to use [metric.SeriesGrouper](https://github.com/influxdata/telegraf/blob/master/metric/series_grouper.go#L24).  Just keep adding tags and fields to this type and then at the end of this function you can range over grouper.Metrics() and call acc.AddMetric().

If I am understanding correctly this will allow you to skip the code to detect tag changes.

Reply:
Done, see https://github.com/influxdata/telegraf/pull/5852/commits/2351ca080ebb02fcbb73cb329fcde10a1de41e73